src/evm/protocol/lido/state.rs

```diff
@@ -0,0 +1,953 @@
+            }),
+        })
+    }
+
+    // amount in is shares
+    fn unwrap_steth(&self, amount_in: BigUint) -> Result<GetAmountOutResult, SimulationError> {
+        if amount_in.is_zero() {
+            return Err(SimulationError::InvalidInput("Cannot unwrap 0 wstETH ".to_string(), None))
+        }
+
+        let amount_out =
+            amount_in.clone() * self.total_pooled_eth.clone() / self.total_shares.clone();
+
+        Ok(GetAmountOutResult {
+            amount: amount_out.clone(),
+            gas: BigUint::from(DEFAULT_GAS),
+            new_state: Box::new(Self {
+                pool_type: self.pool_type.clone(),
+                total_shares: self.total_shares.clone(),
+                total_pooled_eth: self.total_pooled_eth.clone(),
+                total_wrapped_st_eth: Some(
+                    self.total_wrapped_st_eth
+                        .clone()
+                        .unwrap() -
+                        amount_in,
+                ),
+                id: self.id.clone(),
+                native_address: self.native_address.clone(),
+                stake_limits_state: self.stake_limits_state.clone(),
+            }),
+        })
+    }
+
+    fn st_eth_limits(
+        &self,
+        buy_token: Bytes,
+        sell_token: Bytes,
+    ) -> Result<(BigUint, BigUint), SimulationError> {
+        if buy_token == Bytes::from(ST_ETH_ADDRESS_PROXY) && sell_token == Bytes::from(ETH_ADDRESS)
+        {
+            let limit = self.stake_limits_state.get_limit();
+            Ok((limit.clone(), limit))
+        } else if buy_token == Bytes::from(ETH_ADDRESS) &&
+            sell_token == Bytes::from(ST_ETH_ADDRESS_PROXY)
+        {
+            Ok((BigUint::zero(), BigUint::zero()))
+        } else {
+            Err(SimulationError::InvalidInput(
+                format!(
+                    "Get_limits: Invalid combination of tokens for type {:?}: {:?}, {:?}",
+                    self.pool_type, buy_token, sell_token
+                ),
+                None,
+            ))
+        }
+    }
+
+    fn wst_eth_limits(
+        &self,
+        buy_token: Bytes,
+        sell_token: Bytes,
+    ) -> Result<(BigUint, BigUint), SimulationError> {
+        if buy_token == Bytes::from(ST_ETH_ADDRESS_PROXY) &&
+            sell_token == Bytes::from(WST_ETH_ADDRESS)
+        {
+            //amount of wsteth
+            Ok((
+                self.total_wrapped_st_eth
+                    .clone()
+                    .unwrap(),
+                self.total_wrapped_st_eth
+                    .clone()
+                    .unwrap(),
+            ))
+        } else if buy_token == Bytes::from(WST_ETH_ADDRESS) &&
+            sell_token == Bytes::from(ST_ETH_ADDRESS_PROXY)
+        {
+            // total_shares - wstETH
+            let limit_for_wrapping = self.total_shares.clone() -
+                self.total_wrapped_st_eth
+                    .clone()
+                    .unwrap();
```

Comment:
here you are assuming that the limit for wrapping is the total amount of stETH minus the total amount of wstETH that exists right? 🤔